src/Orleans/Cancellation/GrainCancellationToken.cs

```diff
@@ -56,16 +61,19 @@ public sealed class GrainCancellationToken : IDisposable
 
         internal Task Cancel()
         {
-            _cancellationTokenSource.Cancel();
+            // propagate the exception from the _cancellationTokenSource.Cancel back to the caller
+            // but also cancel _targetGrainReferences. 
+            Task task = OrleansTaskExtentions.WrapInTask(_cancellationTokenSource.Cancel);
```

Comment:
instead of this I would just mark the method as async and change the final call to be awaited. Wrapping exceptions in faulted tasks makes the code less readable and could tamper the exceptions details in unexpected ways... You are paying the cost of creating a *new* synthetic Task anyway (via the TCS), so why not just turn the entire method to async? It should have been that way if we assume exceptions could occur.